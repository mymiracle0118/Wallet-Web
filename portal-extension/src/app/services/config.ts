export const walletApiUrl = process.env.BFF_API_ENDPOINT || 'http://localhost:8080'
export const appEnv = process.env.NEXT_PUBLIC_VERCEL_ENV
export const appVersion = process.env.APP_VERSION ?? process.env.NEXT_PUBLIC_APP_VERSION ?? '0.0.0'
