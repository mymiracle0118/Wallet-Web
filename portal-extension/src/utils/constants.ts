const CryptoJS = require('crypto-js')

export const shortenedAddress = (address: string) =>
  [address?.substring(0, 6), address?.substring(address.length - 4, address.length)].join('...')

export const handleCopy = (event: React.MouseEventHandler<HTMLButtonElement>) => {
  // eslint-disable-next-line
  event.stopPropagation()
}

export enum SAFETY_MEASURE {
  HIGH = 'high',
  MEDIUM = 'medium',
  LOW = 'low',
}

export const getWalletAddressRegex = (network: string) => {
  switch (network) {
    case 'APT':
    case 'SUI':
    case 'aptos':
    case 'SUPRA':
      return /^(0x)?[0-9A-Fa-f]{64}$/
    case 'SOL':
      return /^[1-9A-HJ-NP-Za-km-z]{32,}$/
    default:
      return /^0x[a-fA-F0-9]{40}$/
  }
}

export const remove0xStartOfString = (str: string) => {
  if (str.startsWith('0x') || str.startsWith('0X')) {
    str = str.substring(2)
  }
  return str
}

export const encryptData = (data: string, password: string) => {
  const encryptedData = CryptoJS.AES.encrypt(JSON.stringify(data) as string, password).toString()
  return encryptedData
}

export const decryptData = (encryptedData: string, password: string) => {
  const bytes = CryptoJS.AES.decrypt(encryptedData, password)
  const decryptedData = JSON.parse(bytes.toString(CryptoJS.enc.Utf8))
  return decryptedData
}

export enum TRANSACTION {
  COUNT = 15,
  NETWORK_FEE = 100,
  SUPRA_TOKEN_PRICE = 0.05,
}

export const add0xStartOfString = (str: string) => {
  if (!/^0x/i.test(str)) {
    return `0x${str}`
  }
  return str
}
