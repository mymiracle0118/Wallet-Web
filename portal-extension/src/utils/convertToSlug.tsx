export const convertToSlug = (str: string) => {
  return str
    .toLowerCase()
    .replace(/ /g, '-') // Replace spaces with dashes
    .replace(/[^\w-]+/g, '') // Remove non-word characters
}
