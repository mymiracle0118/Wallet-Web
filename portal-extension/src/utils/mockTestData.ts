export const ethereumAddress = '0x63C100ac0C36549C7218070294b60C18d813675c'

export const mockAssets = [
  {
    id: 'ethereum',
    network: 'mainnet',
    chain: 'ethereum',
    type: 'layer1',
    name: 'Ethereum',
    symbol: 'ETH',
    decimal: 18,
    image: 'https://assets.coingecko.com/coins/images/279/large/ethereum.png',
    enabledAsDefault: true,
  },
  {
    id: 'ethereum',
    network: 'goerli',
    chain: 'ethereum',
    type: 'layer1',
    name: 'GoerliETH',
    symbol: 'ETH',
    decimal: 18,
    image: 'https://assets.coingecko.com/coins/images/279/large/ethereum.png',
    enabledAsDefault: false,
  },
]

export const mockAccounts = {
  '0x63C100ac0C36549C7218070294b60C18d813675c': {
    username: 'Account 1',
    address: '0x63C100ac0C36549C7218070294b60C18d813675c',
    assets: [
      {
        chain: 'ethereum',
        decimal: 18,
        enabledAsDefault: true,
        formattedBalance: 0.000307848568706,
        id: 'ethereum',
        image: 'https://assets.coingecko.com/coins/images/279/large/ethereum.png',
        name: 'Ethereum',
        network: 'mainnet',
        symbol: 'ETH',
        type: 'layer1',
      },
      {
        chain: 'ethereum',
        decimal: 18,
        enabledAsDefault: true,
        formattedBalance: 2.034722077599812,
        id: 'ethereum',
        image: 'https://assets.coingecko.com/coins/images/279/large/ethereum.png',
        name: 'GoerliETH',
        network: 'goerli',
        symbol: 'ETH',
        type: 'layer1',
      },
    ],
  },
  '0x1Aa2C1c6799d320c57a0613F1F66Ea9590257bA1': {
    address: '0x1Aa2C1c6799d320c57a0613F1F66Ea9590257bA1',
    username: 'Account2',
    assets: [
      {
        chain: 'ethereum',
        decimal: 18,
        enabledAsDefault: true,
        formattedBalance: 0,
        id: 'ethereum',
        image: 'https://assets.coingecko.com/coins/images/279/large/ethereum.png',
        name: 'Ethereum',
        network: 'mainnet',
        symbol: 'ETH',
        type: 'layer1',
      },
      {
        chain: 'ethereum',
        decimal: 18,
        enabledAsDefault: true,
        formattedBalance: 0,
        id: 'ethereum',
        image: 'https://assets.coingecko.com/coins/images/279/large/ethereum.png',
        name: 'GoerliETH',
        network: 'goerli',
        symbol: 'ETH',
        type: 'layer1',
      },
    ],
  },
}

export const mockAssetsWithSUI = [
  {
    id: 'ethereum',
    network: 'mainnet',
    chain: 'ethereum',
    type: 'layer1',
    name: 'Ethereum',
    symbol: 'ETH',
    decimal: 18,
    image: 'https://assets.coingecko.com/coins/images/279/large/ethereum.png',
    enabledAsDefault: true,
  },
  {
    id: 'ethereum',
    network: 'goerli',
    chain: 'ethereum',
    type: 'layer1',
    name: 'GoerliETH',
    symbol: 'ETH',
    decimal: 18,
    image: 'https://assets.coingecko.com/coins/images/279/large/ethereum.png',
    enabledAsDefault: false,
  },
  {
    id: 'SUI',
    network: 'SUI',
    chain: 'SUI',
    name: 'SUI',
    symbol: 'SUI',
    decimal: 0,
    image: 'https://raw.githubusercontent.com/MystenLabs/sui/main/doc/static/Sui_Icon_Brand.png',
    enabledAsDefault: false,
  },
]

export const mockNetworkAssets = [
  {
    blockExplorer: 'https://goerli.etherscan.io',
    chainId: 5,
    coin: 'Goerli',
    enabledAsDefault: true,
    image: 'https://assets.coingecko.com/coins/images/279/large/ethereum.png',
    name: 'Goerli Test Network',
    networkId: 'goerli',
    symbol: 'GoerliETH',
    testNetwork: true,
  },
]

export const mockTestProfit = [
  {
    network: 'goerli',
    time: 'Mon Sep 26 2022 20:49:21 GMT+0800 (Philippine Standard Time) {}',
    token: 'ETH',
    value: 2.524722077554242,
  },
  {
    network: 'mainnet',
    time: 'Mon Sep 26 2022 20:49:22 GMT+0800 (Philippine Standard Time) {}',
    token: 'ETH',
    value: 0.000307848568706,
  },
]
