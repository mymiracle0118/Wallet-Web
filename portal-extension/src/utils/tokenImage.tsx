import React from 'react'

import ETH from 'assets/coins/ETH.svg'
import USDC from 'assets/coins/USDC.svg'
import SUSHI from 'assets/coins/SUSHI.svg'
import UNI from 'assets/coins/SUP.svg'
import USDT from 'assets/coins/USDT.svg'
import WETH from 'assets/coins/WETH.svg'
import SUI from 'assets/coins/SUI.svg'
import APTOS from 'assets/coins/APTOS.svg'
import SOL from 'assets/coins/SOL.svg'
import SUP from 'assets/coins/SUP.svg'

export const tokenImage = (symbol: string) => {
  switch (symbol) {
    case 'ETH':
      return <ETH />
    case 'USDC':
      return <USDC />
    case 'SUSHI':
      return <SUSHI />
    case 'UNI':
      return <UNI />
    case 'WETH':
      return <WETH />
    case 'USDT':
      return <USDT />
    case 'SUI':
      return <SUI />
    case 'APTOS':
      return <APTOS />
    case 'SOL':
      return <SOL />
    case 'SUPRA':
      return <SUP />
    default:
      return <ETH />
  }
}
