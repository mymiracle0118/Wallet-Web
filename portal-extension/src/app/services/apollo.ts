import { ApolloClient, from, HttpLink, InMemoryCache } from '@apollo/client'
import { onError } from '@apollo/client/link/error'
import { RetryLink } from '@apollo/client/link/retry'
import { ToastConfigs } from '../config/index'
import PATHS from '../../constants/paths'
import { toast } from 'react-toastify'
import { appVersion, walletApiUrl } from './config'
import typeDefs from '../../utils/graphql/typeDefs'

const getClient = () => {
  const accessKey = process.env.API_KEY
  if (!accessKey) {
    throw new Error('API_KEY is not defined in the environment variables.')
  }

  const httpLink = new HttpLink({
    uri: `${walletApiUrl}`,
    fetch,
    headers: {
      Authorization: `Apikey ${accessKey}`,
    },
  })

  const errorLink = onError((error) => {
    const { graphQLErrors = [], networkError } = error
    if (graphQLErrors.length > 0) {
      // eslint-disable-next-line no-console
      const errorMessage = graphQLErrors[0]?.message || 'Internal issue'
      toast.error(errorMessage, ToastConfigs)
      toast.clearWaitingQueue()
    }
    if (networkError) {
      toast.error('Apollo network error, Please check again!', ToastConfigs)
      toast.clearWaitingQueue()
      // eslint-disable-next-line no-console
      console.error(networkError, 'networkError')
      setTimeout(() => {
        localStorage.clear()
        window.location.href = PATHS.API.AUTH.LOGOUT
      }, 1500)
    }
  })

  const mergedLink = from([new RetryLink(), errorLink, httpLink]) //authLink,

  return new ApolloClient({
    ssrMode: false,
    link: mergedLink,
    name: 'pbo-web',
    version: appVersion,
    typeDefs,
    cache: new InMemoryCache({
      typePolicies: {
        LocalState: {
          fields: {},
        },
      },
    }),
  })
}

export default getClient
