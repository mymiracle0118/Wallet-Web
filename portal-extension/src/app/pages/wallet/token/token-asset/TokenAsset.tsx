/* eslint-disable @typescript-eslint/no-floating-promises */
/* eslint-disable @typescript-eslint/restrict-template-expressions */

import React, { useEffect, useMemo, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { useQuery } from '@tanstack/react-query'
import moment from 'moment'
import { NetworkFactory } from '@portal/shared/factory/network.factory'
import classnames from 'classnames'
import SinglePageLayout from 'layouts/single-page-layout/SinglePageLayout'
import { CustomTypography, DataCard, Icon, Loader, TokenActivity, TokenAssetCard } from 'app/components'

import { useWallet } from '@portal/shared/hooks/useWallet'
import { usePricing } from '@portal/shared/hooks/usePricing'
// import { useGas } from '@portal/shared/hooks/useGas'
import { useBalance } from '@portal/shared/hooks/useBalance'
import { ContractTransaction } from '@portal/shared/services/etherscan'

import ExternalLinkIcon from 'assets/icons/external-link.svg'

import { getChartUrl } from '@portal/shared/services/coingecko'
import { createLocationState } from 'lib/woozie'
import { Button, Dropdown, DropdownItem, DropdownMenu, DropdownTrigger, Image } from '@nextui-org/react'
import NoActivity from 'assets/images/no-activity.png'
import {
  CaretDownIcon,
  CheckIcon,
  ReceiveIcon,
  RoundedWhiteIcon,
  SendIcon,
  SignInIcon,
  SwapDownIcon,
  SwapUpIcon,
} from '@src/app/components/Icons'
import { IActivityFilterProps, NetworkToken } from '@portal/shared/utils/types'
import { useStore } from '@portal/shared/hooks/useStore'
import { useGas } from '@portal/shared/hooks/useGas'

const ACTIVITY_FILTER: IActivityFilterProps[] = [
  {
    key: 'All',
    icon: <RoundedWhiteIcon />,
    name: 'All',
  },
  {
    key: 'Send',
    icon: <SendIcon />,
    name: 'Send',
  },
  {
    key: 'Receive',
    icon: <ReceiveIcon />,
    name: 'Receive',
  },
  {
    key: 'Swap_to_other_token',
    icon: <SwapUpIcon />,
    name: 'Swap to other token',
  },
  {
    key: 'Swap_from_other_token',
    icon: <SwapDownIcon />,
    name: 'Swap from other token',
  },
  {
    key: 'Sign',
    icon: <SignInIcon />,
    name: 'Sign',
  },
]

const TokenAsset = () => {
  const { t } = useTranslation()
  const { pathname } = createLocationState()
  const { setPendingTransactions, pendingTransactions, tokenProfitLossCollection } = useWallet()
  const { getMarketPrice, getMarket24HourChange, getAssetValueFormatted } = usePricing()

  const [isTokenFilter, setTokenFilter] = useState(new Set(['All']))
  const [marketGasString, setMarketGasString] = useState('')
  const [pendingTransactionList, setPendingTransactionList] = useState<ContractTransaction[]>([])
  const { getNetworkToken } = useStore()
  // Token Activity Filter
  const filterTokenActivity = useMemo(() => Array.from(isTokenFilter).join(', ').replaceAll('_', ' '), [isTokenFilter])

  const paths = pathname.split('/')
  const network: string = paths[paths.length - 2]
  const assetId: string = paths[paths.length - 1]

  // const asset = useMemo(() => getAsset(network, assetId), [getAsset, network, assetId])
  const asset: NetworkToken = getNetworkToken(assetId)
  const networkAddress = asset?.address
  const enableQuery = Boolean(network && networkAddress && asset)
  const networkFactory = NetworkFactory.selectByNetworkId(asset.networkName)

  const { data: transactions = [], isLoading } = useQuery(
    ['transaction-data', asset.shortName, asset.tokenContractAddress],
    () => networkFactory.getTransactions(network, networkAddress, asset.tokenContractAddress),
    {
      refetchInterval: 6000,
      staleTime: 3000,
      enabled: enableQuery,
    }
  )

  const { balanceFormatted } = useBalance(asset)

  // const assetBalance = useMemo(
  //   () => Math.round((Number(balanceFormatted) + Number.EPSILON) * 10000) / 10000,
  //   [balanceFormatted]
  // )
  // const marketGas = 0
  const { marketGas } = useGas(asset)
  const transactionsWithoutGasTransactions = transactions.filter((v) => {
    return filterTokenActivity.toLowerCase() === 'all'
      ? v.value !== '0'
      : filterTokenActivity.toLowerCase() === 'receive'
      ? v.value !== '0' && v.to.toLowerCase() === networkAddress?.toLowerCase()
      : filterTokenActivity.toLowerCase() === 'send'
      ? v.value !== '0' && v.from.toLowerCase() === networkAddress?.toLowerCase()
      : false
  })
  const comparedValue = useMemo(() => {
    const token = tokenProfitLossCollection.find((v) => v.network === asset.networkName && v.token === asset.shortName)
    return token?.value ? token?.value - balanceFormatted : 0
  }, [balanceFormatted, tokenProfitLossCollection, asset])

  const plValue = useMemo(() => {
    return getAssetValueFormatted(asset.coingeckoTokenId, comparedValue)
  }, [getAssetValueFormatted, asset.coingeckoTokenId, comparedValue])
  const plPercentage = useMemo(
    () => (comparedValue / balanceFormatted) * 100,
    [balanceFormatted, comparedValue]
  ).toFixed(3)

  useEffect(() => {
    if (transactionsWithoutGasTransactions.length > 0 && (pendingTransactions || []).length > 0) {
      const removeNotPendingTransaction: Array<ContractTransaction> = []
      pendingTransactions?.forEach((tx) => {
        const completedTransaction = transactionsWithoutGasTransactions.find((v) => v.hash === tx.hash)
        if (!completedTransaction) {
          removeNotPendingTransaction.push(tx)
        }
      })
      if (removeNotPendingTransaction.length === 0) {
        setPendingTransactions(removeNotPendingTransaction)
      }
      setPendingTransactionList(removeNotPendingTransaction)
    }
  }, [transactionsWithoutGasTransactions]) // eslint-disable-line

  useEffect(() => {
    const fetchMarketGas = async () => {
      const mGasData: string = await marketGas
      setMarketGasString(mGasData)
    }
    fetchMarketGas().then()
  }, [marketGas])

  const selectedIconClass = classnames('absolute right-2 top-4')

  return (
    <SinglePageLayout showMenu BorderBottom={false} paddingClass={false}>
      <div className="px-4 pb-4">
        <div>
          <TokenAssetCard
            image={asset.image}
            dollarAmount={getAssetValueFormatted(asset.coingeckoTokenId, balanceFormatted)}
            coin={asset.title}
            assetId={assetId}
            id={assetId}
            network={network}
            coinAmount={balanceFormatted}
            exchange={asset.tokenGasFeeUnitToDisplay ? `${marketGasString} ${asset.tokenGasFeeUnitToDisplay}` : ''}
            symbol={asset.title}
            isFavorite={asset.isFavorite}
          />

          <div className="my-4 gap-4 justify-between flex items-center">
            <DataCard
              width="50%"
              title={t('Token.tokenPrice', { token: asset.title })}
              amount={`$${getMarketPrice(asset.coingeckoTokenId)}`}
              percentage={getMarket24HourChange(asset.coingeckoTokenId)}
            />
            <DataCard width="50%" title={t('Token.profitLoss')} amount={plValue} percentage={plPercentage} />
          </div>

          {asset && (
            <a
              href={getChartUrl(asset.coingeckoTokenId.toLocaleLowerCase().replace(/\s+/g, '-'))}
              target="_blank"
              rel="noopener noreferrer"
              className="no-underline"
            >
              <DataCard
                title={t('Token.priceChart')}
                icon={
                  <Icon
                    icon={<ExternalLinkIcon className="dark:fill-custom-white fill-custom-black" />}
                    size="medium"
                  />
                }
              />
            </a>
          )}

          <div className="flex items-center justify-between mt-4">
            <CustomTypography type="secondary" variant="subtitle">
              {t('Token.activity')}
            </CustomTypography>

            <Dropdown placement="bottom-end">
              <DropdownTrigger>
                <Button variant="light" className="font-extrabold" size="md">
                  {filterTokenActivity} <CaretDownIcon />
                </Button>
              </DropdownTrigger>
              <DropdownMenu
                aria-label="Token Activity filter"
                variant="flat"
                disallowEmptySelection
                selectionMode="single"
                selectedKeys={isTokenFilter}
                onSelectionChange={setTokenFilter}
                itemClasses={{ selectedIcon: selectedIconClass }}
              >
                {ACTIVITY_FILTER.map((activityItems) => (
                  <DropdownItem
                    key={activityItems.key}
                    startContent={<div className="mr-3">{activityItems.icon}</div>}
                    className="py-3 font-extrabold pr-8"
                    selectedIcon={
                      activityItems.key === isTokenFilter && (
                        <CheckIcon className="w-4 h-4 bg-gradient-primary rounded-full" />
                      )
                    }
                  >
                    {activityItems.name}
                  </DropdownItem>
                ))}
              </DropdownMenu>
            </Dropdown>
          </div>
        </div>

        {pendingTransactionList.map((data) => (
          <div key={data.hash} data-testid={data.hash}>
            <TokenActivity
              page="Token"
              type="Send"
              date={moment.unix(data.time as number).format('MMM. DD [at] hh:mm A')}
              price={`${data.value} ${asset.title}`}
              status={'Pending'}
              transactionHash={data.hash}
              network={network}
              assetId={assetId}
            />
          </div>
        ))}

        {isLoading ? (
          [0, 1, 2].map((id) => (
            <div className="flex px-4 mt-4" key={id}>
              <Loader variant="rounded" className="w-8 h-8" />
              <div className="flex flex-1">
                <div className="flex flex-col w-1/2 pl-5 gap-1">
                  <Loader variant="text" className="!w-9/12" />
                  <Loader variant="text" />
                </div>
                <div className="flex flex-col gap-1 w-1/2 pl-12">
                  <Loader variant="text" />
                  <Loader variant="text" />
                </div>
              </div>
            </div>
          ))
        ) : transactionsWithoutGasTransactions && transactionsWithoutGasTransactions.length > 0 ? (
          transactionsWithoutGasTransactions.map((data) => (
            <div key={data.hash} data-testid={data.hash}>
              <TokenActivity
                transactionHash={data.hash}
                page="Token"
                type={data.from.toLowerCase() === (networkAddress?.toLowerCase() || '') ? 'Send' : 'Received'}
                date={moment.unix(data.time as number).format('MMM. DD [at] hh:mm A')}
                price={`${data.value} ${asset.title}`}
                // status={'Completed'}
                status={data.status && data.status !== 'Invalid' ? 'Completed' : !data.status ? 'Completed' : 'Failed'}
                network={network}
                assetId={assetId}
                tokenDecimal={data.tokenDecimal}
              />
            </div>
          ))
        ) : (
          <div className="flex space-y-3 flex-col items-center mt-4">
            <Image width={160} height={80} src={NoActivity} fallbackSrc={NoActivity} alt="No Activity" />
            <CustomTypography type="secondary" variant="body">
              {t('Token.noActivitiesYet')}
            </CustomTypography>
          </div>
        )}
      </div>
    </SinglePageLayout>
  )
}

export default TokenAsset
