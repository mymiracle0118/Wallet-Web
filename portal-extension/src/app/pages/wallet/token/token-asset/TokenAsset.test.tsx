import React from 'react'
import { render, screen, act, renderHook } from '@testing-library/react'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { I18nextProvider } from 'react-i18next'
import { ethers } from 'ethers'
import { useWallet } from '@portal/shared/hooks/useWallet'

import i18n from '@portal/shared/i18n/config'
import * as etherscan from '@portal/shared/services/etherscan'

import { mockAssets, ethereumAddress } from 'utils/mockTestData'
import TokenAsset from 'pages/wallet/token/token-asset/TokenAsset'

const queryClient = new QueryClient()
const wrapper = (
  <QueryClientProvider client={queryClient}>
    <I18nextProvider i18n={i18n}>
      <TokenAsset />
    </I18nextProvider>
  </QueryClientProvider>
)

const originalLocation = window.location as URL
const mockWindowLocation = (newLocation: URL) => {
  delete window.location
  window.location = newLocation
}
const setLocation = (path: string) => mockWindowLocation(new URL(`https://example.com${path}`))

const mockTransactionData = [
  {
    blockNumber: '7262406',
    time: 'Jul. 20 at 07:42 PM',
    hash: '0xea76c54c02ed88249e4037c0636b85509926b6572b1e250e24d8acac21f70d3c',
    nonce: 24,
    from: '0x63c100ac0c36549c7218070294b60c18d813675c',
    to: '0xeb38b1e8ecb39759704b96b88618ba54ca74d64f',
    value: '0.01',
    gas: ethers.BigNumber.from('0x2386f26fc10000'),
    gasPrice: ethers.BigNumber.from('0x59682f07'),
    gasUsed: ethers.BigNumber.from('0x5208'),
    cumulativeGasUsed: ethers.BigNumber.from('0x041d44'),
  },
]

describe('<TokenAsset />', () => {
  beforeEach(() => {
    const { result: walletResult } = renderHook(() => useWallet())
    walletResult.current.network = 'goerli'
    walletResult.current.assets = mockAssets
    walletResult.current.address = ethereumAddress

    act(() => {
      setLocation('/#/token/goerli/ethereum')
    })

    jest.spyOn(etherscan, 'getTransactions').mockResolvedValue(mockTransactionData)
  })

  afterEach(() => {
    // Restore window.location to not destroy other tests
    mockWindowLocation(originalLocation)
  })

  test('renders correctly', async () => {
    const { asFragment } = render(wrapper)
    await screen.findByTestId(mockTransactionData[0].hash)
    expect(asFragment()).toMatchSnapshot()
  })

  test('should show 1 transaction', async () => {
    render(wrapper)
    expect(await screen.findByTestId(mockTransactionData[0].hash)).toBeInTheDocument()
  })
})
