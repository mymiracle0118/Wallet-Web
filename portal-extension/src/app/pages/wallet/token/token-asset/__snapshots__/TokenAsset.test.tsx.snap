// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TokenAsset /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="dark:bg-surface-dark bg-custom-white flex flex-col h-[600px] dark:text-custom-white text-custom-black"
  >
    <div
      class="flex flex-row relative justify-between items-center h-16"
    >
      <button
        class="h-[24px] w-[24px] p-1 ml-1 hover:bg-custom-grey10 dark:hover:bg-custom-white10 rounded-full flex items-center justify-center"
        type="button"
      >
        <arrow-left.svg
          classname="text-xl stroke-custom-black dark:stroke-custom-white"
        />
      </button>
      <button
        class="profile-btn flex items-center mr-1 text-inherit font-bold hover:bg-custom-grey10 dark:hover:bg-custom-white10 p-3 rounded-md"
        type="button"
      >
        @
        <chevron-right.svg
          classname="stroke-custom-black dark:stroke-custom-white"
        />
      </button>
    </div>
    <div
      class="w-full h-full overflow-auto"
    >
      <div
        class="MuiBox-root css-1cc42ae"
      >
        <div
          class="bg-gradient-bg dark:bg-gradient-bg rounded-lg py-4 px-6"
        >
          <div
            class="relative flex items-center mb-2"
          >
            <img
              alt="token-thumbnail"
              class="h-9 w-9 rounded-full"
              src="https://assets.coingecko.com/coins/images/279/large/ethereum.png"
            />
            <div
              style="flex: 1; margin-left: 16px;"
            >
              <div
                class="break-all text-custom-white text-[1.75rem] font-light"
              >
                0 ETH
              </div>
              <div>
                <div
                  class="text-custom-white text-sm font-extrabold"
                >
                  $0
                </div>
              </div>
            </div>
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="self-start h-[20px] w-[20px] cursor-pointer hover:bg-custom-white10 rounded-full"
              id="headlessui-listbox-button-:r0:"
              type="button"
            >
              <more-horizontal.svg
                classname="text-xl stroke-custom-white"
              />
            </button>
          </div>
          <div
            class="text-custom-white border-[1px] border-solid h-8  border-custom-white40 rounded-xl flex items-center justify-center w-fit px-3"
          >
            <div
              class="text-note text-inherit"
            >
              0x63C1...675c
            </div>
            <div
              class="relative"
            >
              <button
                aria-expanded="false"
                class="ml-1 p-1 focus:outline-none"
                id="headlessui-popover-button-:r2:"
                type="button"
              >
                <copy.svg
                  class="fill-custom-white"
                />
              </button>
            </div>
            <button
              class="ml-0.5 flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full h-[16px] w-[16px] !p-1 !rounded-full dark:hover:bg-custom-white10"
              type="button"
            >
              <external-link.svg
                classname="fill-custom-white"
              />
            </button>
          </div>
          <div
            class="flex gap-3 mt-4 mb-3 items-center justify-between "
          >
            <button
              class="bg-custom-white40 w-4/12 text-custom-white rounded-md p-[12px] max-h-[44px] hover:bg-custom-white hover:text-primary font-bold fill-custom-white hover:fill-primary"
              data-aid="receiveButton"
              type="button"
            >
              Receive
            </button>
            <button
              class="bg-custom-white40 w-4/12 text-custom-white rounded-md p-[12px] max-h-[44px] hover:bg-custom-white hover:text-primary font-bold fill-custom-white hover:fill-primary"
              data-aid="sendButton"
              type="button"
            >
              Send
            </button>
          </div>
          <div
            class="flex items-center"
          >
            <div
              class="asset-text flex-1 ml-2.5 text-custom-white text-sm font-extrabold"
            />
          </div>
        </div>
        <div
          class="MuiBox-root css-hqnnjl"
        >
          <div
            class="flex flex-col p-3 rounded-lg bg-custom-grey10 dark:bg-custom-white10"
            role=""
            style="width: 132px;"
            tabindex="0"
          >
            <div
              class="title-wrapper MuiBox-root css-1lekzkb"
            >
              <div
                class="text-custom-grey100 dark:text-custom-white40 text-sm font-extrabold"
              >
                ETH Price
              </div>
            </div>
            <div
              class="mt-2 mb-2 text-custom-black dark:text-custom-white text-lg font-extrabold"
            >
              0
            </div>
            0
          </div>
          <div
            class="flex flex-col p-3 rounded-lg bg-custom-grey10 dark:bg-custom-white10"
            role=""
            style="width: 132px;"
            tabindex="0"
          >
            <div
              class="title-wrapper MuiBox-root css-1lekzkb"
            >
              <div
                class="text-custom-grey100 dark:text-custom-white40 text-sm font-extrabold"
              >
                Profit Loss
              </div>
            </div>
            <div
              class="mt-2 mb-2 text-custom-black dark:text-custom-white text-lg font-extrabold"
            >
              $0
            </div>
            <div
              class="flex items-center gap-1 text-feedback-negative dark:text-feedback-negative text-sm font-extrabold"
            >
              <svg
                class="inline-block select-none shrink-0 stroke-feedback-positive text-xl"
                style="height: 1em; width: 1em;"
              >
                <trending-down.svg />
              </svg>
              0%
            </div>
          </div>
        </div>
        <a
          href="https://www.coingecko.com/en/coins/goerlieth"
          rel="noopener noreferrer"
          target="_blank"
        >
          <div
            class="flex flex-col p-3 rounded-lg bg-custom-grey10 dark:bg-custom-white10"
            role=""
            style="width: auto;"
            tabindex="0"
          >
            <div
              class="title-wrapper MuiBox-root css-1lekzkb"
            >
              <div
                class="text-custom-grey100 dark:text-custom-white40 text-sm font-extrabold"
              >
                Price Chart
              </div>
              <svg
                class="inline-block select-none shrink-0 fill-custom-black dark:fill-custom-white text-2xl"
                style="height: 1em; width: 1em;"
              >
                <external-link.svg
                  classname="dark:fill-custom-white fill-custom-black"
                />
              </svg>
            </div>
          </div>
        </a>
        <div
          class="MuiBox-root css-69i1ev"
        >
          <div
            class="mt-4 text-custom-grey100 dark:text-custom-white40 text-sm font-extrabold"
          >
            Activity
          </div>
        </div>
      </div>
      <div
        data-testid="0xea76c54c02ed88249e4037c0636b85509926b6572b1e250e24d8acac21f70d3c"
      >
        <div
          class="MuiBox-root css-v18s91"
        >
          <div>
            <svg
              class="inline-block select-none shrink-0 fill-custom-black dark:fill-custom-white text-2xl"
              style="height: 1em; width: 1em;"
            >
              <arrow-up-right.svg />
            </svg>
          </div>
          <div
            style="margin-left: 28px; flex: 1;"
          >
            <div
              class="text-h4 text-custom-black dark:text-custom-white"
            >
              Send
            </div>
            <div
              class="text-subtitle1 text-custom-grey100 dark:text-custom-white40"
            >
              Invalid date
            </div>
          </div>
          <div
            style="text-align: right;"
          >
            <div
              class="text-h4 text-custom-black dark:text-custom-white"
            >
              0.01 ETH
            </div>
            <div
              class="text-subtitle1 text-custom-black dark:text-custom-white"
            >
              Completed
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
